### PR TITLE
Implement Y-flow create feature 

### PR DESCRIPTION
**Pull request by sergii-iakovenko**
The changes:
- Implement the create operation for Y-flow (this doesn't include installation of y-point meter yet).
- Integrate with PCE y-point finding method.
- Add blank functional tests for y-flow CRUD.

**MERGE NOTE:** merge after #4530 and change the base correspondingly.

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/YFlowCreateService.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    /**
+     * Handles async response from worker.
+     *
+     * @param key command identifier.
+     */
+    public void handleAsyncResponse(String key, SpeakerResponse flowResponse) {
+        log.debug("Received flow command response: {}", flowResponse);
+        YFlowCreateFsm fsm = fsms.get(key);
+        if (fsm == null) {
+            log.debug("Can't find a FSM: received response with key {} for non pending FSM", key);
```
Comment (surabujin) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/YFlowCreateService.java:183]: In most cases this is error marker, on "debug" level we will never see it. Even when it is not an error I would like to see this message on prod, to be able to measure actual delay between speaker request and speaker response.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/YFlowValidator.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        if (subFlows.size() < 2) {
+            throw new InvalidFlowException(
+                    format("The y-flow '%s' must have at least 2 sub flows", request.getYFlowId()),
+                    ErrorType.DATA_INVALID);
+        }
+
+        for (SubFlowDto subFlow : subFlows) {
+            if (request.getYFlowId() == null) {
```
Comment (surabujin) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/YFlowValidator.java:82]: Should it be `subFlow.getFlowId()` ?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/YFlowValidator.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                throw new InvalidFlowException(
+                        format("The sub-flow of y-flow '%s' has no sub-flow id provided", request.getYFlowId()),
+                        ErrorType.DATA_INVALID);
+            }
+            if (subFlow.getSharedEndpoint() == null) {
+                throw new InvalidFlowException(
+                        format("The sub-flow of y-flow '%s' has no shared endpoint provided", request.getYFlowId()),
```
Comment (surabujin) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/YFlowValidator.java:89]: Good to have sub flow id somewhere in this message.

File: src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/InMemoryPathComputer.java [hunk 4]
```diff
@@ -298,4 +304,62 @@ private Path convertToPath(SwitchId srcSwitchId, SwitchId dstSwitchId, List<Edge
...
+                throw new IllegalArgumentException(format("The path '%s' has no path segments", flowPath.getPathId()));
+            }
+
+            LinkedList<SwitchId> path = new LinkedList<>();
+            path.add(pathSegments.get(0).getSrcSwitchId());
+            for (PathSegment pathSegment : pathSegments) {
+                path.add(pathSegment.getDestSwitchId());
+            }
+
+            if (sharedSwitchId.equals(path.getLast())) {
```
Comment (surabujin) [inline src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/InMemoryPathComputer.java:352]: Unexpected action, method name do not supposed anything like this.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/create/action/OnReceivedInstallResponseAction.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+        if (response.isSuccess()) {
+            stateMachine.removePendingCommand(commandId);
+            stateMachine.saveActionToHistory("Rule was installed",
+                    format("The rule was installed: switch %s", response.getSwitchId()));
+        } else {
+            FlowErrorResponse errorResponse = (FlowErrorResponse) response;
+
+            int attempt = stateMachine.doRetryForCommand(commandId);
+            if (attempt <= speakerCommandRetriesLimit) {
```
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/create/action/OnReceivedInstallResponseAction.java:60]: And as result each caller must do this "check".

**Review by surabujin — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/bolts/YFlowCreateHubBolt.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        yFlowCreateService.activate();
+    }
+
+
+    @Override
+    protected void onRequest(Tuple input) throws PipelineException {
+        currentKey = pullKey(input);
+        YFlowRequest payload = pullValue(input, FIELD_ID_PAYLOAD, YFlowRequest.class);
+        try {
+            yFlowCreateService.handleRequest(currentKey, pullContext(input), payload);
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/bolts/YFlowCreateHubBolt.java:146]: Make sense. Replaced.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java [hunk 7]
```diff
@@ -490,8 +437,26 @@ protected String getCrudActionName() {
                     .addEntryAction(new OnFinishedWithErrorAction(dashboardLogger));
         }
 
-        public FlowCreateFsm produce(String flowId, CommandContext commandContext) {
-            FlowCreateFsm fsm = builder.newStateMachine(State.INITIALIZED, flowId, commandContext, carrier, config);
+        public FlowCreateFsm newInstance(CommandContext commandContext, String flowId, boolean allowNorthboundResponse,
+                                         @NonNull Collection<FlowCreateEventListener> eventListeners) {
+            FlowCreateFsm fsm = builder.newStateMachine(State.INITIALIZED, commandContext, carrier, flowId,
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java:442]: What is the point? AutoStart is enabled. This FSM starts with the context built outside the method, not null/empty context.

I'd prefer to start FSM only after it's properly placed into the system (FSM map, etc.)

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/action/ResourcesAllocationAction.java [hunk 8]
```diff
@@ -385,4 +389,12 @@ private void updateSwitchRelatedFlowProperties(Flow flow) {
         }
         flow.setDetectConnectedDevices(detectConnectedDevices.build());
     }
+
+    @Override
+    protected void handleError(FlowCreateFsm stateMachine, Exception ex, ErrorType errorType, boolean logTraceback) {
+        super.handleError(stateMachine, ex, errorType, logTraceback);
+
+        // Notify about failed allocation.
+        stateMachine.notifyEventListenersOnError(errorType, stateMachine.getErrorReason());
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/action/ResourcesAllocationAction.java:398]: Correct. It's done for reasons: pass the error type to event listeners (YFlowCreate) and notify ASAP.  I don't see a problem with 2 error events.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/create/action/OnReceivedInstallResponseAction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/yflow/create/action/OnReceivedInstallResponseAction.java:60]: I don't see any benefits from moving this check into the method. Each caller needs the retry result (the attempt is less than speakerCommandRetriesLimit) + the attempt number for error message. So the check will be still here.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (surabujin) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/create/FlowCreateFsm.java:442]: In `netwrok` it was required to call `enter` methods on initial state. It bring several "surprise" incidents, so I have made a "rule" for myself to call this `.start()` method unconditionally.

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/model/system/FeatureTogglesDto.java [hunk 9]
```diff
@@ -62,6 +62,9 @@
     @JsonProperty("server42_isl_rtt")
     private Boolean server42IslRtt;
 
+    @JsonProperty("cud_on_yflow_enabled")
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/model/system/FeatureTogglesDto.java]: CUD is intentionally here, as reading is always allowed.

File: docker/db-migration/migrations/017-add-yflow-classes.yaml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        - sql: "DROP CLASS yflow"
+  - changeSet:
+      id: y_subflow_class
+      author: siakovenko
+      changes:
+        - sql: "CREATE CLASS y_subflow IF NOT EXISTS EXTENDS E"
+        - sql: "CREATE PROPERTY y_subflow.yflow_id IF NOT EXISTS STRING"
+        - sql: "CREATE PROPERTY y_subflow.subflow_id IF NOT EXISTS STRING"
+        - sql: "CREATE INDEX y_subflow.yflow_id NOTUNIQUE_HASH_INDEX"
+        - sql: "CREATE INDEX y_subflow.subflow_id NOTUNIQUE_HASH_INDEX"
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply docker/db-migration/migrations/017-add-yflow-classes.yaml]: Make sense. Changed to unique.

File: src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/FlowResourcesManager.java [hunk 11]
```diff
@@ -237,4 +237,23 @@ public void deallocateCookie(long unmaskedCookie) {
...
+
+    /**
+     * Try to allocate meter which is bound to a flow only.
+     * The method doesn't initialize a transaction. So it requires external transaction to cover allocation failures.
+     */
+    public MeterId allocateMeter(String flowId, SwitchId switchId) throws ResourceAllocationException {
+        try {
+            return meterPool.allocate(switchId, flowId, null);
+        } catch (ConstraintViolationException | ResourceNotAvailableException ex) {
+            throw new ResourceAllocationException("Unable to allocate meter", ex);
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/flow/resources/FlowResourcesManager.java]: Revised, although it's not required - ResourceNotAvailableException carries the details.

File: src-java/kilda-model/src/main/java/org/openkilda/model/YFlow.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+
+        @Mapping(target = "subFlows", ignore = true)
+        void copyWithoutSubFlows(YFlowData source, @MappingTarget YFlowData target);
+
+        /**
+         * Performs deep copy of entity data.
+         *
+         * @param source the y-flow to copy from.
+         */
+        default YFlowData deepCopy(YFlowData source, YFlow targetFlow) {
```
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply src-java/kilda-model/src/main/java/org/openkilda/model/YFlow.java]: Discussed several times. We definitely don't want the cloner having a state. No need for access modifiers here. Interface declaration is just less verbose in this case.

**Pull request merged**